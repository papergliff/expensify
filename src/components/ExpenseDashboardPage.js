import React from 'react'

const ExpenseDashboardPage = () => (
  <div>
    <p>This is from my dashboard component</p> 
  </div>
)

export default ExpenseDashboardPage 