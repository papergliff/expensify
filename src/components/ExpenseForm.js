import React from 'react'
import moment from 'moment'
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates'

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props)
    this.onDescriptionChange = this.onDescriptionChange.bind(this)
    this.onNoteChange = this.onNoteChange.bind(this)
    this.onAmountChange = this.onAmountChange.bind(this)
    this.onDateChange = this.onDateChange.bind(this)
    this.onFocusChange = this.onFocusChange.bind(this)
    this.onSubmit = this.onSubmit.bind(this)

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: '',
    }
  }

  onDescriptionChange(e) {
    const description = e.target.value
    this.setState(() => ({ description }))
  }

  onNoteChange(e) {
    const note = e.target.value
    this.setState(() => ({ note }))
  }

  onAmountChange(e) {
    const amount = e.target.value

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  }

  onDateChange(createdAt) {
    if (createdAt) {
      this.setState(() => ({ createdAt }))
    }
  }

  onFocusChange({ focused })  {
    this.setState(() => ({ calendarFocused: focused }))
  }

  onSubmit(e) {
    e.preventDefault()
    const {
      description,
      amount,
      createdAt,
      note,
    } = this.state
    const { onSubmit } = this.props

    if (!description || !amount) {
      this.setState(() => ({ error: 'Please provide description and amount' }))
    } else {
      this.setState(() => ({ error: '' }))
      onSubmit({
        description,
        amount: parseFloat(amount, 10) * 100,
        createdAt: createdAt.valueOf(),
        note,
      })
    }
  }

  render() {
    const {
      error,
      description,
      amount,
      createdAt,
      calendarFocused,
      note,
    } = this.state
    return (
      <div>
        {error && <p>{error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            value={description}
            onChange={this.onDescriptionChange}
            autoFocus
          />
          <input
            type="text"
            placeholder="Amount"
            value={amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={createdAt}
            onDateChange={this.onDateChange}
            focused={calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={note}
            onChange={this.onNoteChange}
          />
          <button>Add Expense</button>
        </form>
      </div>
    )
  }
}
