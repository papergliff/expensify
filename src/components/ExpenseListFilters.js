import React from 'react'
import { connect } from 'react-redux'
import { setTextFilter } from '../actions/filters'

const ExpenseListFilters = props => (
  <div>
    <input type="text" defaultValue={props.filters.text} onChange={(e) => {
      props.dispatch(setTextFilter(e.target.value))
      console.log(e.target.value)
    }} />
  </div>
)

const mapStateToProps = state => ({
  filters: state.filters
})

export default connect(mapStateToProps)(ExpenseListFilters)