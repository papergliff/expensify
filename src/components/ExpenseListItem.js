import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import moment from 'moment'
import numeral from 'numeral'
import { removeExpense } from '../actions/expenses'

export const ExpenseListItem = ({ id, description, amount, createdAt, dispatch }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>
      {numeral(amount / 100).format('$0,0.00')}
      -
      {moment(createdAt).format('MMMM Do, YYYY')}
    </p>
    <button onClick={() => dispatch(removeExpense({ id }))}>Remove</button>
  </div>
)

export default connect()(ExpenseListItem)
