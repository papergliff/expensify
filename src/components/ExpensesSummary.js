import React from 'react'
import { connect } from 'react-redux'
import numeral from 'numeral'
import selectExpenses from '../selectors/expenses'
import getExpensesTotal from '../selectors/expenses-total'

export const ExpensesSummary = ({ expenses, expensesTotal }) => {
  const expenseWord = expenses.length === 1 ? 'expense' : 'expenses'
  const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00')
  return (
    <div>
      <h1>Viewing {expenses.length} {expenseWord} totalling {formattedExpensesTotal}</h1>
    </div>
  )
}

const mapStateToProps = state => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters)

  return {
    expenses: visibleExpenses,
    expensesTotal: getExpensesTotal(visibleExpenses),
  }
}

export default connect(mapStateToProps)(ExpensesSummary)
