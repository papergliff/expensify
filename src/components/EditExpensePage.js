import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { editExpense, removeExpense } from '../actions/expenses'

export class EditExpensePage extends React.Component {
  constructor(props) {
    super(props)

    this.onSubmit = this.onSubmit.bind(this)
    this.onRemove = this.onRemove.bind(this)
  }

  onSubmit(expense) {
    const { editExpense, history } = this.props

    editExpense(expense)
    history.push('/')
  }

  onRemove() {
    const { removeExpense, history } = this.props

    removeExpense()
    history.push('/')
  }

  render() {
    const { expense } = this.props

    return (
      <div>
        <ExpenseForm
          expense={expense}
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: expense => dispatch(editExpense(props.match.params.id, expense)),
  removeExpense: () => dispatch(removeExpense({ id: props.match.params.id }))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)