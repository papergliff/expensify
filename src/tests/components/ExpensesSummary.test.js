import React from 'react'
import { shallow } from 'enzyme'
import { ExpensesSummary } from '../../components/ExpensesSummary'
import getExpensesTotal from '../../selectors/expenses-total'
import expenses from '../fixtures/expenses'

test('should render expenses summary component with one expense', () => {
  const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]} expensesTotal={getExpensesTotal([expenses[0]])} />)
  expect(wrapper).toMatchSnapshot()
})

test('should render expenses summary component with multiple expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenses={expenses} expensesTotal={getExpensesTotal(expenses)} />)
  expect(wrapper).toMatchSnapshot()
})