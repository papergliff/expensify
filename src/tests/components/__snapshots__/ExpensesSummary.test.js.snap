// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expenses summary component with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    3
     
    expenses
     totalling 
    $1,141.95
  </h1>
</div>
`;

exports[`should render expenses summary component with one expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $1.95
  </h1>
</div>
`;
